File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.cpp
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        // vector<int> result;
        // for(int i = 0; i<spells.size();i++){
        //     int count = 0;
        //     for(int j = 0; j<potions.size();j++){
        //         if(spells[i]*potions[j]>=success) count++;
        //     }
        //     result.push_back(count);
        // }
        // return result;

        sort(potions.begin(), potions.end());
        int m = potions.size();
        vector<int> ans(spells.size());
        
        for (int i = 0; i < (int)spells.size(); ++i) {
            long long s = spells[i];
            long long need = (success + s - 1) / s; 
            auto it = lower_bound(potions.begin(), potions.end(), need);
            ans[i] = (int)(potions.end() - it);
        }
        return ans;
    }
};

File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        Arrays.sort(potions);
        int n = spells.length;
        int m = potions.length;
        int[] result = new int[n];
        
        for (int i = 0; i < n; i++) {
            long s = spells[i];
            long need = (success + s - 1) / s; 

            int left = 0, right = m - 1, pos = m;
            while (left <= right) {
                int mid = left + (right - left) / 2;
                if (potions[mid] >= need) {
                    pos = mid;
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
            result[i] = m - pos;
        }
        
        return result;
    }
}
